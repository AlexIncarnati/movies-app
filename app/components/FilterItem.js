import React from 'react';
import Proptypes from 'prop-types';

const FilterItem = (props) => {
	const genre = props.genre.name;
	const genreId = props.genre.id;
	const checked = props.genre.visible;

	return (
		<div className="filter-genres__item pretty p-default p-curve p-thick p-smooth">
			<input
				type="checkbox"
				className="filter-genres__checkbox"
				defaultValue={genre}
				name={genre}
				id={genreId}
				defaultChecked={checked}
				onClick={props.handleChangeFilter}
			/>
			<div className="state p-danger-o">
				<label htmlFor={genre} className="filter-genres__checkbox-label">
					{genre}
				</label>
			</div>
		</div>
	);
};

FilterItem.propTypes = {
	genre: Proptypes.object,
	genreId: Proptypes.string,
	checked: Proptypes.bool
};

export default FilterItem;
