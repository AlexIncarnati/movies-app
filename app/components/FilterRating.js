import React from 'react';
import Proptypes from 'prop-types';

const FilterRating = (props) => {
	return (
		<select onChange={props.handleChangeRatings}>
			<option defaultValue="1">1</option>
			<option defaultValue="2">2</option>
			<option defaultValue="3">3</option>
			<option defaultValue="4">4</option>
			<option defaultValue="5">5</option>
			<option defaultValue="6">6</option>
			<option defaultValue="7">7</option>
			<option defaultValue="8">8</option>
			<option defaultValue="9">9</option>
			<option defaultValue="10">10</option>
		</select>
	);
};

FilterRating.propTypes = {
	handleChangeRatings: Proptypes.func
};

export default FilterRating;
