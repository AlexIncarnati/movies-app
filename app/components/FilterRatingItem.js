import React from 'react';

const FilterRatingItem = (props) => {
	const value = props.value;
	return <option value={value}>{value}</option>;
};

export default FilterRatingItem;
