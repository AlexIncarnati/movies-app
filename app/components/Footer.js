import React from 'react';

const Footer = () => {
	return (
		<footer>
			<h5 className="footer-title">
				<a href="http://www.rightinterfaces.com" target="_blank" rel="noopener noreferrer">
					Alessandro Incarnati Â© 2018
				</a>
			</h5>
		</footer>
	);
};

export default Footer;
