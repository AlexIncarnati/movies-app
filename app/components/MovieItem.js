import React from 'react';
import Proptypes from 'prop-types';
import displayplaceholderImage from '../images/placeholder.png';

const MovieItem = props => {
  const {
    movie: { vote_average, title, poster_path, popularity },
  } = props;
  const { movie } = props;
  let { genres, genresVisibles } = props;
  genres = genres.join(', ');
  const movieDbImagePath = 'https://image.tmdb.org/t/p/w500/';
  const posterPath = poster_path ? `${movieDbImagePath}${poster_path}` : displayplaceholderImage;
  genresVisibles = genresVisibles.every(item => item === true);

  if (genresVisibles) {
    return (
      <div className="movie-item">
        <img src={posterPath} alt={title} />
        <div className="movie-item__wrapper">
          <h5 className="movie-item__title">{title}</h5>
          <p className="movie-item__genre">
            <span>Genre</span> {genres}
          </p>
          <p className="movie-item__rating">
            <span>Rating</span> {vote_average}
          </p>
          <p className="movie-item__popularity">
            <span>Popularity</span> {popularity}
          </p>
        </div>
      </div>
    );
  }
  return null;
};

MovieItem.propTypes = {
  movie: Proptypes.object,
  genres: Proptypes.array,
  genresVisibles: Proptypes.array,
};

export default MovieItem;
