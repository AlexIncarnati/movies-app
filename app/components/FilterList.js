import React, { Component } from 'react';
import FilterItem from './FilterItem';
import Proptypes from 'prop-types';

class FilterList extends Component {
	constructor(props) {
		super(props);
		this.state = { checked: this.props.visible };
	}
	render() {
		return (
			<div className="filter-genres-wrap">
				<h3 className="filter-genres__title">Filter by genres</h3>
				<div className="filter-genres">
					{this.props.genres ? (
						this.props.genres.map((genre, index) => (
							<FilterItem
								key={index}
								genre={genre}
								handleChangeFilter={this.props.handleChangeFilter}
								visible={genre.visible}
								checked={this.state.checked}
							/>
						))
					) : (
						''
					)}
				</div>
			</div>
		);
	}
}

FilterList.propTypes = {
	genre: Proptypes.object,
	handleChangeFilter: Proptypes.func,
	checked: Proptypes.bool,
	visible: Proptypes.bool
};

export default FilterList;
