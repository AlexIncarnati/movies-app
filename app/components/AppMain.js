import React, { Component } from 'react';
import axios from 'axios';
import Loading from './Loading';
import { config } from '../config/config';
import Header from '../components/Header';
import MovieList from './MovieList';
import FilterList from './FilterList';
import FilterRating from './FilterRating';
import Footer from './Footer';

// Define variables
let loaderTimeout;
const landingMovieId = config.LANDING_MOVIE_ID;
let url = `${config.BASE_URL}/movie/now_playing?&api_key=${config.KEY}&language=en-US&page=1`;
let urlGenres = `${config.BASE_URL}/genre/movie/list?&api_key=${config.KEY}&language=en-US`;

class MoviesMain extends Component {
	constructor(props) {
		super(props);

		this.state = {
			moviesData: [],
			moviesGenres: [],
			allMoviesData: [],
			loading: true
		};

		this.handleChangeFilter = this.handleChangeFilter.bind(this);
		this.handleChangeRatings = this.handleChangeRatings.bind(this);
	}

	handleChangeFilter(event) {
		const target = event.target;
		const value = target.name;
		// Convert into number to be able to compare it by type
		const id = parseFloat(target.id);
		let moviesGenres = this.state.moviesGenres;

		moviesGenres.forEach((item) => {
			if (item.id === id) {
				item.visible = !item.visible;
			}
		});

		this.setState({
			moviesGenres: moviesGenres
		});
	}

	handleChangeRatings(event) {
		const target = event.target;
		const value = parseFloat(target.value);
		// Reload all movies data
		const storedMoviesData = this.state.allMoviesData;
		let moviesData = JSON.parse(JSON.stringify(storedMoviesData));
		// Filter by vote average
		moviesData = moviesData.filter((item) => item.vote_average >= value);

		this.setState({
			moviesData: moviesData
		});
	}

	setGenreVisibility(data) {
		data.forEach((item) => (item.visible = true));
		this.setState({
			moviesGenres: data
		});
		return data;
	}

	// Get all movies data from API using fetch and save into state
	getMoviesData() {
		axios
			.get(url)
			.then((data) =>
				// handle success
				this.setState({
					moviesData: data.data.results,
					allMoviesData: data.data.results
				})
			)
			.catch((error) => {
				// handle error
				this.setState({ error: error });
			});
	}

	// Get all movies data from API using fetch and save into state
	getMoviesGenres() {
		let dataGenres;
		axios
			.get(urlGenres)
			// handle success
			.then((data) => {
				dataGenres = data.data.genres;
				dataGenres = this.setGenreVisibility(dataGenres);
				this.setState({
					moviesGenres: dataGenres
				});
			})
			.catch((error) => {
				this.setState({ error: error });
			});
	}

	componentWillMount() {
		// Api Calls
		this.getMoviesData();
		this.getMoviesGenres();
		clearTimeout(loaderTimeout);
	}

	// Setting state of loader function
	setLoadingState() {
		this.setState({ loading: false });
	}

	// Set timeout after all elements of Component have loaded
	componentDidMount() {
		loaderTimeout = setTimeout(() => {
			this.setLoadingState();
		}, 4000);
	}

	render() {
		const loading = this.state.loading;
		const error = this.state.error;
		const movies = this.state.moviesData;

		if (loading) {
			return <Loading />;
		}
		if (error) {
			return (
				<div className="main">
					<Header />
					<h3 className="no-results">There has been some issue fetching the API data</h3>
					<p>{this.state.error}</p>
					<Footer />
				</div>
			);
		}
		if (!movies.length) {
			return (
				<div className="main">
					<Header />
					<FilterList genres={this.state.moviesGenres} handleChangeFilter={this.handleChangeFilter} />
					<FilterRating handleChangeRatings={this.handleChangeRatings} />
					<h3 className="no-results">No movies found matching your criteria.</h3>
					<Footer />
				</div>
			);
		}
		return (
			<div className="main">
				<Header />
				<FilterList genres={this.state.moviesGenres} handleChangeFilter={this.handleChangeFilter} />
				<FilterRating handleChangeRatings={this.handleChangeRatings} />
				<MovieList movies={this.state.moviesData} genres={this.state.moviesGenres} />
				<Footer />
			</div>
		);
	}
}

export default MoviesMain;
