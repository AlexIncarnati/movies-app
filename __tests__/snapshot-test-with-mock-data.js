import React from 'react';
import AppMain from '../app/components/AppMain';
import FilterItem from '../app/components/FilterItem';
import FilterList from '../app/components/FilterList';
import FilterRating from '../app/components/FilterRating';
import FilterRatingItem from '../app/components/FilterRating';
import Header from '../app/components/Header';
import Footer from '../app/components/Footer';
import MovieItem from '../app/components/MovieItem';

import { shallow } from 'enzyme';

describe('The AppMain Component', () => {
	it('should render correctly', () => {
		const component = shallow(<AppMain />);

		expect(component).toMatchSnapshot();
	});
});

describe('The FilterItem Component', () => {
	it('should render correctly', () => {
		const genre = {
			id: 28,
			name: 'Action',
			visible: true
		};
		const component = shallow(<FilterItem genre={{ genre }} />);

		expect(component).toMatchSnapshot();
	});
});

describe('The FilterList Component', () => {
	it('should render correctly', () => {
		const genres = [
			{
				id: 28,
				name: 'Action',
				visible: true
			},
			{
				id: 12,
				name: 'Adventure',
				visible: true
			},
			{
				id: 16,
				name: 'Animation',
				visible: true
			},
			{
				id: 35,
				name: 'Comedy',
				visible: true
			},
			{
				id: 80,
				name: 'Crime',
				visible: true
			},
			{
				id: 99,
				name: 'Documentary',
				visible: true
			},
			{
				id: 18,
				name: 'Drama',
				visible: true
			},
			{
				id: 10751,
				name: 'Family',
				visible: true
			},
			{
				id: 14,
				name: 'Fantasy',
				visible: true
			},
			{
				id: 36,
				name: 'History',
				visible: true
			},
			{
				id: 27,
				name: 'Horror',
				visible: true
			},
			{
				id: 10402,
				name: 'Music',
				visible: true
			},
			{
				id: 9648,
				name: 'Mystery',
				visible: true
			},
			{
				id: 10749,
				name: 'Romance',
				visible: true
			},
			{
				id: 878,
				name: 'Science Fiction',
				visible: true
			},
			{
				id: 10770,
				name: 'TV Movie',
				visible: true
			},
			{
				id: 53,
				name: 'Thriller',
				visible: true
			},
			{
				id: 10752,
				name: 'War',
				visible: true
			},
			{
				id: 37,
				name: 'Western',
				visible: true
			}
		];
		const component = shallow(<FilterList props={{ genres }} />);

		expect(component).toMatchSnapshot();
	});
});

describe('The FilterRating Component', () => {
	it('should render correctly', () => {
		const component = shallow(<FilterRating />);

		expect(component).toMatchSnapshot();
	});
});

describe('The FilterRatingItem Component', () => {
	it('should render correctly', () => {
		const component = shallow(<FilterRatingItem />);

		expect(component).toMatchSnapshot();
	});
});

describe('The Footer Component', () => {
	it('should render correctly', () => {
		const component = shallow(<Footer />);

		expect(component).toMatchSnapshot();
	});
});

describe('The Header Component', () => {
	it('should render correctly', () => {
		const component = shallow(<Header />);

		expect(component).toMatchSnapshot();
	});
});

describe('The MovieItem Component', () => {
	it('should render correctly', () => {
		const movie = {
			vote_average: 6.6,
			title: 'Venom',
			genre_ids: [ 878, 28 ],
			poster_path: '/2uNW4WbgBXL25BAbXGLnLqX71Sw.jpg',
			popularity: 432.604
		};
		const genres = [ 'Science Fiction', 'Action' ];
		const genresVisibles = [ true, true ];
		const component = shallow(<MovieItem movie={{ movie }} genres={genres} genresVisibles={genresVisibles} />);

		expect(component).toMatchSnapshot();
	});
});
