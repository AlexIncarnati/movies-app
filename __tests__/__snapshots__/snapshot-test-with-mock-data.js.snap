// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AppMain Component should render correctly 1`] = `<Loading />`;

exports[`The FilterItem Component should render correctly 1`] = `
<div
  className="filter-genres__item pretty p-default p-curve p-thick p-smooth"
>
  <input
    className="filter-genres__checkbox"
    type="checkbox"
  />
  <div
    className="state p-danger-o"
  >
    <label
      className="filter-genres__checkbox-label"
    />
  </div>
</div>
`;

exports[`The FilterList Component should render correctly 1`] = `
<div
  className="filter-genres"
>
  <h3
    className="filter-genres__title"
  >
    Filter by genres
  </h3>
  <div
    className="filter-genres__menu"
  />
</div>
`;

exports[`The FilterRating Component should render correctly 1`] = `
<div
  className="filter-rating"
>
  <h3
    className="filter-rating__title"
  >
    Filter by rating
  </h3>
  <div
    className="filter-rating__select"
  >
    <select
      defaultValue="3"
    >
      <FilterRatingItem
        key="0"
        value={0}
      />
      <FilterRatingItem
        key="1"
        value={0.5}
      />
      <FilterRatingItem
        key="2"
        value={1}
      />
      <FilterRatingItem
        key="3"
        value={1.5}
      />
      <FilterRatingItem
        key="4"
        value={2}
      />
      <FilterRatingItem
        key="5"
        value={2.5}
      />
      <FilterRatingItem
        key="6"
        value={3}
      />
      <FilterRatingItem
        key="7"
        value={3.5}
      />
      <FilterRatingItem
        key="8"
        value={4}
      />
      <FilterRatingItem
        key="9"
        value={4.5}
      />
      <FilterRatingItem
        key="10"
        value={5}
      />
      <FilterRatingItem
        key="11"
        value={5.5}
      />
      <FilterRatingItem
        key="12"
        value={6}
      />
      <FilterRatingItem
        key="13"
        value={6.5}
      />
      <FilterRatingItem
        key="14"
        value={7}
      />
      <FilterRatingItem
        key="15"
        value={7.5}
      />
      <FilterRatingItem
        key="16"
        value={8}
      />
      <FilterRatingItem
        key="17"
        value={8.5}
      />
      <FilterRatingItem
        key="18"
        value={9}
      />
      <FilterRatingItem
        key="19"
        value={9.5}
      />
      <FilterRatingItem
        key="20"
        value={10}
      />
    </select>
  </div>
</div>
`;

exports[`The FilterRatingItem Component should render correctly 1`] = `
<div
  className="filter-rating"
>
  <h3
    className="filter-rating__title"
  >
    Filter by rating
  </h3>
  <div
    className="filter-rating__select"
  >
    <select
      defaultValue="3"
    >
      <FilterRatingItem
        key="0"
        value={0}
      />
      <FilterRatingItem
        key="1"
        value={0.5}
      />
      <FilterRatingItem
        key="2"
        value={1}
      />
      <FilterRatingItem
        key="3"
        value={1.5}
      />
      <FilterRatingItem
        key="4"
        value={2}
      />
      <FilterRatingItem
        key="5"
        value={2.5}
      />
      <FilterRatingItem
        key="6"
        value={3}
      />
      <FilterRatingItem
        key="7"
        value={3.5}
      />
      <FilterRatingItem
        key="8"
        value={4}
      />
      <FilterRatingItem
        key="9"
        value={4.5}
      />
      <FilterRatingItem
        key="10"
        value={5}
      />
      <FilterRatingItem
        key="11"
        value={5.5}
      />
      <FilterRatingItem
        key="12"
        value={6}
      />
      <FilterRatingItem
        key="13"
        value={6.5}
      />
      <FilterRatingItem
        key="14"
        value={7}
      />
      <FilterRatingItem
        key="15"
        value={7.5}
      />
      <FilterRatingItem
        key="16"
        value={8}
      />
      <FilterRatingItem
        key="17"
        value={8.5}
      />
      <FilterRatingItem
        key="18"
        value={9}
      />
      <FilterRatingItem
        key="19"
        value={9.5}
      />
      <FilterRatingItem
        key="20"
        value={10}
      />
    </select>
  </div>
</div>
`;

exports[`The Footer Component should render correctly 1`] = `
<footer>
  <h5
    className="footer-title"
  >
    <a
      href="http://www.rightinterfaces.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Alessandro Incarnati Â© 2018
    </a>
  </h5>
</footer>
`;

exports[`The Header Component should render correctly 1`] = `
<header>
  <svg
    data-name="Layer 1"
    fill="#fff"
    viewBox="0 0 251.34 222.81"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      The Movie Db Logo
    </title>
    <path
      d="M2944.56,2236.34c14.6,0,24.48-9.88,24.48-24.48V2052.58c0-14.6-9.88-24.48-24.48-24.48H2742.19c-14.6,0-24.48,9.88-24.48,24.48v198.33l12.56-14.56h0V2052.58a11.94,11.94,0,0,1,11.92-11.92h202.37a11.94,11.94,0,0,1,11.92,11.92v159.28a11.94,11.94,0,0,1-11.92,11.92H2760.82l-12.56,12.56-0.08-.1"
      transform="translate(-2717.71 -2028.09)"
    />
    <polygon
      points="61.38 77.29 68.32 77.29 68.32 49.52 76.96 49.52 76.96 42.63 52.74 42.63 52.74 49.52 61.38 49.52 61.38 77.29"
    />
    <polygon
      points="99.53 77.29 106.47 77.29 106.47 42.58 99.53 42.58 99.53 56.46 89.14 56.46 89.14 42.58 82.2 42.58 82.2 77.29 89.14 77.29 89.14 63.4 99.53 63.4 99.53 77.29"
    />
    <polygon
      points="132.25 70.34 119.23 70.34 119.23 63.4 130.35 63.4 130.35 56.46 119.23 56.46 119.23 49.52 131.66 49.52 131.66 42.58 112.28 42.58 112.28 77.29 132.25 77.29 132.25 70.34"
    />
    <polygon
      points="68.66 101.35 54.97 86.11 52.74 86.11 52.74 121.35 59.78 121.35 59.78 101.98 68.66 111.3 77.54 101.98 77.5 121.35 84.53 121.35 84.53 86.11 82.35 86.11 68.66 101.35"
    />
    <path
      d="M2825.75,2114.16c-23.88,0-23.88,35.77,0,35.77S2849.63,2114.16,2825.75,2114.16Zm0,28.59c-13.88,0-13.88-21.45,0-21.45S2839.63,2142.75,2825.75,2142.75Z"
      transform="translate(-2717.71 -2028.09)"
    />
    <rect
      height="34.7"
      width="6.94"
      x="165.37"
      y="86.65"
    />
    <polygon
      points="185.07 114.41 185.07 107.47 196.19 107.47 196.19 100.53 185.07 100.53 185.07 93.59 197.5 93.59 197.5 86.65 178.13 86.65 178.13 121.35 198.09 121.35 198.09 114.41 185.07 114.41"
    />
    <path
      d="M2780.79,2158.81h-10.34v34.7h10.34C2803.89,2193.51,2803.89,2158.81,2780.79,2158.81Zm0,27.76h-3.4v-20.82h3.4C2794.28,2165.75,2794.28,2186.57,2780.79,2186.57Z"
      transform="translate(-2717.71 -2028.09)"
    />
    <path
      d="M2824,2176.13c2.18-1.5,3.11-4.22,3.2-6.84,0.15-6.12-3.69-10.53-9.85-10.53h-13.74v34.75h13.74a10.32,10.32,0,0,0,10.24-10.44A8.43,8.43,0,0,0,2824,2176.13Zm-13.4-10.44h6.17a3.51,3.51,0,0,1,0,7h-6.17v-7Zm6.17,20.87h-6.17v-6.94h6.17a3.41,3.41,0,0,1,3.49,3.45A3.45,3.45,0,0,1,2816.77,2186.57Z"
      transform="translate(-2717.71 -2028.09)"
    />
    <polygon
      points="144.01 105.38 134.87 86.65 126.86 86.65 143.23 122.08 144.79 122.08 161.15 86.65 153.14 86.65 144.01 105.38"
    />
  </svg>
  <h1
    className="header-title"
  >
    Top 20 Movies out now
  </h1>
</header>
`;

exports[`The MovieItem Component should render correctly 1`] = `
<div
  className="movie-item"
>
  <img
    src={null}
  />
  <div
    className="movie-item__wrapper"
  >
    <h5
      className="movie-item__title"
    />
    <p
      className="movie-item__genre"
    >
      <span>
        Genre
      </span>
       
      Science Fiction, Action
    </p>
    <p
      className="movie-item__rating"
    >
      <span>
        Rating
      </span>
       
    </p>
    <p
      className="movie-item__popularity"
    >
      <span>
        Popularity
      </span>
       
    </p>
  </div>
</div>
`;
